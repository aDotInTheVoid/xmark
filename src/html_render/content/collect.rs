
use std::ffi::OsStr;
use std::path::{Path, PathBuf};

use eyre::Result;

use crate::config::GlobalConf;
use crate::{cli, config, summary};

use super::{Book, Content, Link, Page};