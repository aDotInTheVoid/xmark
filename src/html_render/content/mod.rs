
use eyre::Result;
use serde::{Deserialize, Serialize};
use std::ffi::OsStr;
use std::path::{Path, PathBuf};

pub mod collect;
pub mod pagetoc;
pub use collect::Dirs;



#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct Page {
    pub name: String,
    /// The html file to render to
    pub output: PathBuf,
    /// The md input file.
    pub input: PathBuf,
    pub toc: (),
    /// The link to the next page
    pub next: Option<String>,
    /// The link to the previous
    pub prev: Option<String>,
    /// The pages parents, and their parents, and so on.
    ///
    /// Inspired by [github's docs](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github/set-up-git)
    /// Where is has the heirachy "[GitHub.com](https://docs.github.com/en/free-pro-team@latest/github) /
    /// [Getting started](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github) /
    /// [Quickstart](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github/quickstart) /
    /// [Set up Git](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github/set-up-git)
    pub heirachy: ()
}
// Oh dear god the allocations

/// Fun helper type
///
/// 1. Foo
/// 2. Bar
/// 2.1. Baz
/// 2.1.1 Quix
/// 2.2 Spam
///
/// Chapter(Foo)
/// Chapter(Bar)
/// StartSection
/// Chapter(Baz)
/// StartSection
enum PageListParts<'a> {
    //TODO: A better name
    Chapter(&'a ()),
    StartSection,
    EndSection,
}
