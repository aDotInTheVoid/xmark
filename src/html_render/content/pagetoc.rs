//! The "minitoc" for the page
//!
//! This is inspired by the right hand table of contents from
//! [MkDocs Material](https://squidfunk.github.io/mkdocs-material/getting-started/)
//!
//! It's different from the left hand table for contents, which is for the whole book.
//! This is just for the the page we're on
//!
//! A page with this content
//! ```markdown
//! # Controll Flow
//!
//! ## If
//! ### If else
//! ### As an expression
//!
//! ## Match
//!
//! ## Loops
//! ### for .. in ..
//! ### while ...
//! ### loop
//! ```
//!
//! Will have this as a toc
//! ```rust,notest
//! # https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=14a9939a1d620cd67a9c8763a730a6df
//! PageToc(
//!    [
//!        H2 {
//!            this: Link {
//!                pritty: "If",
//!                link: "if",
//!            },
//!            children: [
//!                H3(
//!                    Link {
//!                        pritty: "If else",
//!                        link: "if-else",
//!                    },
//!                ),
//!                H3(
//!                    Link {
//!                        pritty: "As an expression",
//!                        link: "as-as-expression",
//!                    },
//!                ),
//!            ],
//!        },
//!        H2 {
//!            this: Link {
//!                pritty: "Match",
//!                link: "match",
//!            },
//!            children: [],
//!        },
//!        H2 {
//!            this: Link {
//!                pritty: "Loops",
//!                link: "loops",
//!            },
//!            children: [
//!                H3(
//!                    Link {
//!                        pritty: "for .. in ..",
//!                        link: "for-in",
//!                    },
//!                ),
//!                H3(
//!                    Link {
//!                        pritty: "while ...",
//!                        link: "while",
//!                    },
//!                ),
//!                H3(
//!                    Link {
//!                        pritty: "loop",
//!                        link: "loop",
//!                    },
//!                ),
//!            ],
//!        },
//!    ],
//!)
//!```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct PageToc(pub Vec<H2>);

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct Link {
    // The "nice" name, eg "Creating a book"
    pub pritty: String,
    // The name of the link, eg "creating-a-book"
    pub link: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct H3(pub Link);

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct H2 {
    pub this: Link,
    pub children: Vec<H3>,
}
