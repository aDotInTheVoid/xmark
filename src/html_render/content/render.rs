use std::fs;

use eyre::Result;
use pulldown_cmark::{html, Options, Parser};
use ramhorns::Content as Rhc;
use serde::Serialize;

use super::{Link, Page as CPage};

// Because we borrow link, we cant Deserialize, so snapshot tests may not work.
// If so, we can remove the Serialize bound
#[derive(Debug, Clone, Serialize, PartialEq, Rhc)]
pub struct Page<'a> {
    title: &'a str,
    inner_html: String,
    pub heirachy: &'a [Link],
    pub pagetoc: String,
    /// The link to the next page
    pub next: Option<&'a str>,
    /// The link to the previous
    pub prev: Option<&'a str>,
    #[ramhorns(flatten)]
    pub global: Global<'a>,
}

impl<'a> Page<'a> {
    pub fn new(from: &'a CPage, dirs: &'a super::Dirs) -> Result<Self> {
        // TODO: Don't buffer the whole input
        let inner_html = render_markdown(&fs::read_to_string(&from.input)?);

        let global = Global {
            path_to_root: &dirs.base_url,
            ..Default::default()
        };

        Ok(Self {
            title: &from.name,
            inner_html,
            // TODO: The way to do this is to gen the pagetoc in render_markdown.
            pagetoc: String::new(),
            next: from.next.as_deref(),
            prev: from.prev.as_deref(),
            heirachy: &from.heirachy,
            global,
        })
    }
}

/// Options every page needs not specific to a page
#[derive(Debug, Clone, Serialize, PartialEq, Rhc)]
pub struct Global<'a> {
    pub path_to_root: &'a str,
    pub language: &'a str,
    pub preferred_dark_theme: &'a str,
    pub default_theme: &'a str
}

impl<'a> Default for Global<'a> {
    fn default() -> Self {
        Global {
            path_to_root: "/",
            language: "en",
            // THESE ARE FACTS.
            default_theme: "rust",
            preferred_dark_theme: "coal"
        }
    }
}

// TODO: A million customizations
pub(crate) fn render_markdown(content: &str) -> String {
    let opts = Options::all();
    let parser = Parser::new_ext(content, opts);
    let mut out = String::new();
    html::push_html(&mut out, parser);
    out
}
