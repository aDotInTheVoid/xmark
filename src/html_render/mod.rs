// SPDX-License-Identifier: GPL-3.0-only
use std::fmt::Debug;
use std::fs;
use std::io::Write;

use eyre::{Context, Result};
use fs_extra::dir as fsx;
use ramhorns::{Content as Rhc, Ramhorns};
use tracing::instrument;

use crate::cli;
use crate::cli::config::GlobalConf;

use crate::content::{self, Book, Content, Page};

mod tpl_types;


/// Singleton
pub(crate) struct HTMLRender<'a> {
    content: Content,
    // I'll need em later, when this gets fancy
    _args: &'a cli::Args,
    templates: Ramhorns,
    pub(crate) dirs: content::Dirs,
}

impl Debug for HTMLRender<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("HTMLRender")
            .field("dirs", &self.dirs)
            .field("content", &self.content)
            .finish()
    }
}

impl<'a> HTMLRender<'a> {
    #[instrument]
    pub(crate) fn new(conf: &GlobalConf, args: &'a cli::Args) -> Result<Self> {
        let dirs = content::Dirs::new(conf, args);

        //TODO: This wount work for incrmental or multi-renderer
        //TODO: Embed static
        if dirs.out_dir.exists() {
            fs::remove_dir_all(&dirs.out_dir)?;
        }

        let parent = dirs.out_dir.parent().unwrap();

        fs::create_dir_all(parent)?;

        fsx::copy(
            concat!(env!("CARGO_MANIFEST_DIR"), "/www/static/"),
            parent,
            &Default::default(),
        )?;

        fs::rename(parent.join("static"), &dirs.out_dir)?;

        let content = content::Content::new(conf, &dirs)?;

        let templates =
            Ramhorns::from_folder(concat!(env!("CARGO_MANIFEST_DIR"), "/www/templates/"))?;

        Ok(Self {
            content,
            _args: args,
            templates,
            dirs,
        })
    }
    
    #[instrument]
    pub(crate) fn render(&self) -> Result<()> {
        //TODO: Rayon
        for book in &self.content.0 {
            for page in &book.pages {
                let html = self.render_page(page, book)?;
                fs::create_dir_all(page.output.parent().unwrap())?;
                let mut file = fs::File::create(&page.output)
                    .wrap_err_with(|| format!("Failed to create {:?}", &page.output))?;
                file.write_all(html.as_bytes())?;
            }

            for (file, url) in &book.redirects {
                #[derive(Rhc)]
                struct Params<'a> {
                    url: &'a str,
                }

                self.templates
                    .get("redirect.html")
                    .unwrap()
                    .render_to_file(file, &Params { url })?;
            }
        }

        Ok(())
    }

    #[instrument]
    pub(crate) fn render_page(&self, page: &Page, book: &Book) -> Result<String> {
        let rp = tpl_types::Page::new(page, &self, book)?;
        let tpl = self.templates.get("page.html").unwrap();
        // TODO: Use render_to_file or something
        Ok(tpl.render(&rp))
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::{
        cli::{self, config},
        html_render,
    };

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "book-*/**"],
        )
        .unwrap();

        let args = cli::Args {
            dir: temp.path().to_owned(),
            ..Default::default()
        };
        let conf = config::load(&args).unwrap();
        let render = html_render::HTMLRender::new(&conf, &args).unwrap();
        render.render().unwrap();

        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }

    // #[test]
    // fn render_readmes() {
    //     glob!("render_html_tests/*.md", |path| {
    //         let input = fs::read_to_string(path).unwrap();
    //         let out = render_chap(&input);
    //         assert_snapshot!(out);
    //     })
    // }
}
