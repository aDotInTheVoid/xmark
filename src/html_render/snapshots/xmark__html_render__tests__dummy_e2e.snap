---
source: src/html_render/mod.rs
expression: paths
---
- ""
- /_out
- /_out/html
- /_out/html/book-1
- /_out/html/book-1/appendix-00
- /_out/html/book-1/appendix-00/index.html
- /_out/html/book-1/appendix-01-keywords
- /_out/html/book-1/appendix-01-keywords/index.html
- /_out/html/book-1/appendix-02-operators
- /_out/html/book-1/appendix-02-operators/index.html
- /_out/html/book-1/appendix-03-derivable-traits
- /_out/html/book-1/appendix-03-derivable-traits/index.html
- /_out/html/book-1/appendix-04-useful-development-tools
- /_out/html/book-1/appendix-04-useful-development-tools/index.html
- /_out/html/book-1/appendix-05-editions
- /_out/html/book-1/appendix-05-editions/index.html
- /_out/html/book-1/appendix-06-translation
- /_out/html/book-1/appendix-06-translation/index.html
- /_out/html/book-1/appendix-07-nightly-rust
- /_out/html/book-1/appendix-07-nightly-rust/index.html
- /_out/html/book-1/ch00-00-introduction
- /_out/html/book-1/ch00-00-introduction/index.html
- /_out/html/book-1/ch01-00-getting-started
- /_out/html/book-1/ch01-00-getting-started/index.html
- /_out/html/book-1/ch01-01-installation
- /_out/html/book-1/ch01-01-installation/index.html
- /_out/html/book-1/ch01-02-hello-world
- /_out/html/book-1/ch01-02-hello-world/index.html
- /_out/html/book-1/ch01-03-hello-cargo
- /_out/html/book-1/ch01-03-hello-cargo/index.html
- /_out/html/book-1/ch02-00-guessing-game-tutorial
- /_out/html/book-1/ch02-00-guessing-game-tutorial/index.html
- /_out/html/book-1/ch03-00-common-programming-concepts
- /_out/html/book-1/ch03-00-common-programming-concepts/index.html
- /_out/html/book-1/ch03-01-variables-and-mutability
- /_out/html/book-1/ch03-01-variables-and-mutability/index.html
- /_out/html/book-1/ch03-02-data-types
- /_out/html/book-1/ch03-02-data-types/index.html
- /_out/html/book-1/ch03-03-how-functions-work
- /_out/html/book-1/ch03-03-how-functions-work/index.html
- /_out/html/book-1/ch03-04-comments
- /_out/html/book-1/ch03-04-comments/index.html
- /_out/html/book-1/ch03-05-control-flow
- /_out/html/book-1/ch03-05-control-flow/index.html
- /_out/html/book-1/ch04-00-understanding-ownership
- /_out/html/book-1/ch04-00-understanding-ownership/index.html
- /_out/html/book-1/ch04-01-what-is-ownership
- /_out/html/book-1/ch04-01-what-is-ownership/index.html
- /_out/html/book-1/ch04-02-references-and-borrowing
- /_out/html/book-1/ch04-02-references-and-borrowing/index.html
- /_out/html/book-1/ch04-03-slices
- /_out/html/book-1/ch04-03-slices/index.html
- /_out/html/book-1/ch05-00-structs
- /_out/html/book-1/ch05-00-structs/index.html
- /_out/html/book-1/ch05-01-defining-structs
- /_out/html/book-1/ch05-01-defining-structs/index.html
- /_out/html/book-1/ch05-02-example-structs
- /_out/html/book-1/ch05-02-example-structs/index.html
- /_out/html/book-1/ch05-03-method-syntax
- /_out/html/book-1/ch05-03-method-syntax/index.html
- /_out/html/book-1/ch06-00-enums
- /_out/html/book-1/ch06-00-enums/index.html
- /_out/html/book-1/ch06-01-defining-an-enum
- /_out/html/book-1/ch06-01-defining-an-enum/index.html
- /_out/html/book-1/ch06-02-match
- /_out/html/book-1/ch06-02-match/index.html
- /_out/html/book-1/ch06-03-if-let
- /_out/html/book-1/ch06-03-if-let/index.html
- /_out/html/book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules
- /_out/html/book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules/index.html
- /_out/html/book-1/ch07-01-packages-and-crates
- /_out/html/book-1/ch07-01-packages-and-crates/index.html
- /_out/html/book-1/ch07-02-defining-modules-to-control-scope-and-privacy
- /_out/html/book-1/ch07-02-defining-modules-to-control-scope-and-privacy/index.html
- /_out/html/book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree
- /_out/html/book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree/index.html
- /_out/html/book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword
- /_out/html/book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword/index.html
- /_out/html/book-1/ch07-05-separating-modules-into-different-files
- /_out/html/book-1/ch07-05-separating-modules-into-different-files/index.html
- /_out/html/book-1/ch08-00-common-collections
- /_out/html/book-1/ch08-00-common-collections/index.html
- /_out/html/book-1/ch08-01-vectors
- /_out/html/book-1/ch08-01-vectors/index.html
- /_out/html/book-1/ch08-02-strings
- /_out/html/book-1/ch08-02-strings/index.html
- /_out/html/book-1/ch08-03-hash-maps
- /_out/html/book-1/ch08-03-hash-maps/index.html
- /_out/html/book-1/ch09-00-error-handling
- /_out/html/book-1/ch09-00-error-handling/index.html
- /_out/html/book-1/ch09-01-unrecoverable-errors-with-panic
- /_out/html/book-1/ch09-01-unrecoverable-errors-with-panic/index.html
- /_out/html/book-1/ch09-02-recoverable-errors-with-result
- /_out/html/book-1/ch09-02-recoverable-errors-with-result/index.html
- /_out/html/book-1/ch09-03-to-panic-or-not-to-panic
- /_out/html/book-1/ch09-03-to-panic-or-not-to-panic/index.html
- /_out/html/book-1/ch10-00-generics
- /_out/html/book-1/ch10-00-generics/index.html
- /_out/html/book-1/ch10-01-syntax
- /_out/html/book-1/ch10-01-syntax/index.html
- /_out/html/book-1/ch10-02-traits
- /_out/html/book-1/ch10-02-traits/index.html
- /_out/html/book-1/ch10-03-lifetime-syntax
- /_out/html/book-1/ch10-03-lifetime-syntax/index.html
- /_out/html/book-1/ch11-00-testing
- /_out/html/book-1/ch11-00-testing/index.html
- /_out/html/book-1/ch11-01-writing-tests
- /_out/html/book-1/ch11-01-writing-tests/index.html
- /_out/html/book-1/ch11-02-running-tests
- /_out/html/book-1/ch11-02-running-tests/index.html
- /_out/html/book-1/ch11-03-test-organization
- /_out/html/book-1/ch11-03-test-organization/index.html
- /_out/html/book-1/ch12-00-an-io-project
- /_out/html/book-1/ch12-00-an-io-project/index.html
- /_out/html/book-1/ch12-01-accepting-command-line-arguments
- /_out/html/book-1/ch12-01-accepting-command-line-arguments/index.html
- /_out/html/book-1/ch12-02-reading-a-file
- /_out/html/book-1/ch12-02-reading-a-file/index.html
- /_out/html/book-1/ch12-03-improving-error-handling-and-modularity
- /_out/html/book-1/ch12-03-improving-error-handling-and-modularity/index.html
- /_out/html/book-1/ch12-04-testing-the-librarys-functionality
- /_out/html/book-1/ch12-04-testing-the-librarys-functionality/index.html
- /_out/html/book-1/ch12-05-working-with-environment-variables
- /_out/html/book-1/ch12-05-working-with-environment-variables/index.html
- /_out/html/book-1/ch12-06-writing-to-stderr-instead-of-stdout
- /_out/html/book-1/ch12-06-writing-to-stderr-instead-of-stdout/index.html
- /_out/html/book-1/ch13-00-functional-features
- /_out/html/book-1/ch13-00-functional-features/index.html
- /_out/html/book-1/ch13-01-closures
- /_out/html/book-1/ch13-01-closures/index.html
- /_out/html/book-1/ch13-02-iterators
- /_out/html/book-1/ch13-02-iterators/index.html
- /_out/html/book-1/ch13-03-improving-our-io-project
- /_out/html/book-1/ch13-03-improving-our-io-project/index.html
- /_out/html/book-1/ch13-04-performance
- /_out/html/book-1/ch13-04-performance/index.html
- /_out/html/book-1/ch14-00-more-about-cargo
- /_out/html/book-1/ch14-00-more-about-cargo/index.html
- /_out/html/book-1/ch14-01-release-profiles
- /_out/html/book-1/ch14-01-release-profiles/index.html
- /_out/html/book-1/ch14-02-publishing-to-crates-io
- /_out/html/book-1/ch14-02-publishing-to-crates-io/index.html
- /_out/html/book-1/ch14-03-cargo-workspaces
- /_out/html/book-1/ch14-03-cargo-workspaces/index.html
- /_out/html/book-1/ch14-04-installing-binaries
- /_out/html/book-1/ch14-04-installing-binaries/index.html
- /_out/html/book-1/ch14-05-extending-cargo
- /_out/html/book-1/ch14-05-extending-cargo/index.html
- /_out/html/book-1/ch15-00-smart-pointers
- /_out/html/book-1/ch15-00-smart-pointers/index.html
- /_out/html/book-1/ch15-01-box
- /_out/html/book-1/ch15-01-box/index.html
- /_out/html/book-1/ch15-02-deref
- /_out/html/book-1/ch15-02-deref/index.html
- /_out/html/book-1/ch15-03-drop
- /_out/html/book-1/ch15-03-drop/index.html
- /_out/html/book-1/ch15-04-rc
- /_out/html/book-1/ch15-04-rc/index.html
- /_out/html/book-1/ch15-05-interior-mutability
- /_out/html/book-1/ch15-05-interior-mutability/index.html
- /_out/html/book-1/ch15-06-reference-cycles
- /_out/html/book-1/ch15-06-reference-cycles/index.html
- /_out/html/book-1/ch16-00-concurrency
- /_out/html/book-1/ch16-00-concurrency/index.html
- /_out/html/book-1/ch16-01-threads
- /_out/html/book-1/ch16-01-threads/index.html
- /_out/html/book-1/ch16-02-message-passing
- /_out/html/book-1/ch16-02-message-passing/index.html
- /_out/html/book-1/ch16-03-shared-state
- /_out/html/book-1/ch16-03-shared-state/index.html
- /_out/html/book-1/ch16-04-extensible-concurrency-sync-and-send
- /_out/html/book-1/ch16-04-extensible-concurrency-sync-and-send/index.html
- /_out/html/book-1/ch17-00-oop
- /_out/html/book-1/ch17-00-oop/index.html
- /_out/html/book-1/ch17-01-what-is-oo
- /_out/html/book-1/ch17-01-what-is-oo/index.html
- /_out/html/book-1/ch17-02-trait-objects
- /_out/html/book-1/ch17-02-trait-objects/index.html
- /_out/html/book-1/ch17-03-oo-design-patterns
- /_out/html/book-1/ch17-03-oo-design-patterns/index.html
- /_out/html/book-1/ch18-00-patterns
- /_out/html/book-1/ch18-00-patterns/index.html
- /_out/html/book-1/ch18-01-all-the-places-for-patterns
- /_out/html/book-1/ch18-01-all-the-places-for-patterns/index.html
- /_out/html/book-1/ch18-02-refutability
- /_out/html/book-1/ch18-02-refutability/index.html
- /_out/html/book-1/ch18-03-pattern-syntax
- /_out/html/book-1/ch18-03-pattern-syntax/index.html
- /_out/html/book-1/ch19-00-advanced-features
- /_out/html/book-1/ch19-00-advanced-features/index.html
- /_out/html/book-1/ch19-01-unsafe-rust
- /_out/html/book-1/ch19-01-unsafe-rust/index.html
- /_out/html/book-1/ch19-03-advanced-traits
- /_out/html/book-1/ch19-03-advanced-traits/index.html
- /_out/html/book-1/ch19-04-advanced-types
- /_out/html/book-1/ch19-04-advanced-types/index.html
- /_out/html/book-1/ch19-05-advanced-functions-and-closures
- /_out/html/book-1/ch19-05-advanced-functions-and-closures/index.html
- /_out/html/book-1/ch19-06-macros
- /_out/html/book-1/ch19-06-macros/index.html
- /_out/html/book-1/ch20-00-final-project-a-web-server
- /_out/html/book-1/ch20-00-final-project-a-web-server/index.html
- /_out/html/book-1/ch20-01-single-threaded
- /_out/html/book-1/ch20-01-single-threaded/index.html
- /_out/html/book-1/ch20-02-multithreaded
- /_out/html/book-1/ch20-02-multithreaded/index.html
- /_out/html/book-1/ch20-03-graceful-shutdown-and-cleanup
- /_out/html/book-1/ch20-03-graceful-shutdown-and-cleanup/index.html
- /_out/html/book-1/foreword
- /_out/html/book-1/foreword/index.html
- /_out/html/book-1/title-page
- /_out/html/book-1/title-page/index.html
- /_out/html/book-2
- /_out/html/book-2/cli
- /_out/html/book-2/cli/build
- /_out/html/book-2/cli/build/index.html
- /_out/html/book-2/cli/clean
- /_out/html/book-2/cli/clean/index.html
- /_out/html/book-2/cli/index.html
- /_out/html/book-2/cli/init
- /_out/html/book-2/cli/init/index.html
- /_out/html/book-2/cli/serve
- /_out/html/book-2/cli/serve/index.html
- /_out/html/book-2/cli/test
- /_out/html/book-2/cli/test/index.html
- /_out/html/book-2/cli/watch
- /_out/html/book-2/cli/watch/index.html
- /_out/html/book-2/continuous-integration
- /_out/html/book-2/continuous-integration/index.html
- /_out/html/book-2/for_developers
- /_out/html/book-2/for_developers/backends
- /_out/html/book-2/for_developers/backends/index.html
- /_out/html/book-2/for_developers/index.html
- /_out/html/book-2/for_developers/preprocessors
- /_out/html/book-2/for_developers/preprocessors/index.html
- /_out/html/book-2/format
- /_out/html/book-2/format/config
- /_out/html/book-2/format/config/index.html
- /_out/html/book-2/format/index.html
- /_out/html/book-2/format/mathjax
- /_out/html/book-2/format/mathjax/index.html
- /_out/html/book-2/format/mdbook
- /_out/html/book-2/format/mdbook/index.html
- /_out/html/book-2/format/summary
- /_out/html/book-2/format/summary/index.html
- /_out/html/book-2/format/theme
- /_out/html/book-2/format/theme/editor
- /_out/html/book-2/format/theme/editor/index.html
- /_out/html/book-2/format/theme/index-hbs
- /_out/html/book-2/format/theme/index-hbs/index.html
- /_out/html/book-2/format/theme/index.html
- /_out/html/book-2/format/theme/syntax-highlighting
- /_out/html/book-2/format/theme/syntax-highlighting/index.html
- /_out/html/book-2/index.html
- /_out/html/book-2/misc
- /_out/html/book-2/misc/contributors
- /_out/html/book-2/misc/contributors/index.html
- /_out/html/book-3
- /_out/html/book-3/c1
- /_out/html/book-3/c1/index.html
- /_out/html/book-3/c2
- /_out/html/book-3/c2/index.html
- /_out/html/book-3/c3
- /_out/html/book-3/c3/index.html
- /_out/html/book-3/c31
- /_out/html/book-3/c31/index.html
- /_out/html/book-3/c32
- /_out/html/book-3/c32/index.html
- /_out/html/book-3/c33
- /_out/html/book-3/c33/index.html
- /_out/html/book-3/post1
- /_out/html/book-3/post1/index.html
- /_out/html/book-3/pre1
- /_out/html/book-3/pre1/index.html
- /_out/html/book-3/pre2
- /_out/html/book-3/pre2/index.html
- /book-1
- /book-1/SUMMARY.md
- /book-1/appendix-00.md
- /book-1/appendix-01-keywords.md
- /book-1/appendix-02-operators.md
- /book-1/appendix-03-derivable-traits.md
- /book-1/appendix-04-useful-development-tools.md
- /book-1/appendix-05-editions.md
- /book-1/appendix-06-translation.md
- /book-1/appendix-07-nightly-rust.md
- /book-1/ch00-00-introduction.md
- /book-1/ch01-00-getting-started.md
- /book-1/ch01-01-installation.md
- /book-1/ch01-02-hello-world.md
- /book-1/ch01-03-hello-cargo.md
- /book-1/ch02-00-guessing-game-tutorial.md
- /book-1/ch03-00-common-programming-concepts.md
- /book-1/ch03-01-variables-and-mutability.md
- /book-1/ch03-02-data-types.md
- /book-1/ch03-03-how-functions-work.md
- /book-1/ch03-04-comments.md
- /book-1/ch03-05-control-flow.md
- /book-1/ch04-00-understanding-ownership.md
- /book-1/ch04-01-what-is-ownership.md
- /book-1/ch04-02-references-and-borrowing.md
- /book-1/ch04-03-slices.md
- /book-1/ch05-00-structs.md
- /book-1/ch05-01-defining-structs.md
- /book-1/ch05-02-example-structs.md
- /book-1/ch05-03-method-syntax.md
- /book-1/ch06-00-enums.md
- /book-1/ch06-01-defining-an-enum.md
- /book-1/ch06-02-match.md
- /book-1/ch06-03-if-let.md
- /book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules.md
- /book-1/ch07-01-packages-and-crates.md
- /book-1/ch07-02-defining-modules-to-control-scope-and-privacy.md
- /book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.md
- /book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword.md
- /book-1/ch07-05-separating-modules-into-different-files.md
- /book-1/ch08-00-common-collections.md
- /book-1/ch08-01-vectors.md
- /book-1/ch08-02-strings.md
- /book-1/ch08-03-hash-maps.md
- /book-1/ch09-00-error-handling.md
- /book-1/ch09-01-unrecoverable-errors-with-panic.md
- /book-1/ch09-02-recoverable-errors-with-result.md
- /book-1/ch09-03-to-panic-or-not-to-panic.md
- /book-1/ch10-00-generics.md
- /book-1/ch10-01-syntax.md
- /book-1/ch10-02-traits.md
- /book-1/ch10-03-lifetime-syntax.md
- /book-1/ch11-00-testing.md
- /book-1/ch11-01-writing-tests.md
- /book-1/ch11-02-running-tests.md
- /book-1/ch11-03-test-organization.md
- /book-1/ch12-00-an-io-project.md
- /book-1/ch12-01-accepting-command-line-arguments.md
- /book-1/ch12-02-reading-a-file.md
- /book-1/ch12-03-improving-error-handling-and-modularity.md
- /book-1/ch12-04-testing-the-librarys-functionality.md
- /book-1/ch12-05-working-with-environment-variables.md
- /book-1/ch12-06-writing-to-stderr-instead-of-stdout.md
- /book-1/ch13-00-functional-features.md
- /book-1/ch13-01-closures.md
- /book-1/ch13-02-iterators.md
- /book-1/ch13-03-improving-our-io-project.md
- /book-1/ch13-04-performance.md
- /book-1/ch14-00-more-about-cargo.md
- /book-1/ch14-01-release-profiles.md
- /book-1/ch14-02-publishing-to-crates-io.md
- /book-1/ch14-03-cargo-workspaces.md
- /book-1/ch14-04-installing-binaries.md
- /book-1/ch14-05-extending-cargo.md
- /book-1/ch15-00-smart-pointers.md
- /book-1/ch15-01-box.md
- /book-1/ch15-02-deref.md
- /book-1/ch15-03-drop.md
- /book-1/ch15-04-rc.md
- /book-1/ch15-05-interior-mutability.md
- /book-1/ch15-06-reference-cycles.md
- /book-1/ch16-00-concurrency.md
- /book-1/ch16-01-threads.md
- /book-1/ch16-02-message-passing.md
- /book-1/ch16-03-shared-state.md
- /book-1/ch16-04-extensible-concurrency-sync-and-send.md
- /book-1/ch17-00-oop.md
- /book-1/ch17-01-what-is-oo.md
- /book-1/ch17-02-trait-objects.md
- /book-1/ch17-03-oo-design-patterns.md
- /book-1/ch18-00-patterns.md
- /book-1/ch18-01-all-the-places-for-patterns.md
- /book-1/ch18-02-refutability.md
- /book-1/ch18-03-pattern-syntax.md
- /book-1/ch19-00-advanced-features.md
- /book-1/ch19-01-unsafe-rust.md
- /book-1/ch19-03-advanced-traits.md
- /book-1/ch19-04-advanced-types.md
- /book-1/ch19-05-advanced-functions-and-closures.md
- /book-1/ch19-06-macros.md
- /book-1/ch20-00-final-project-a-web-server.md
- /book-1/ch20-01-single-threaded.md
- /book-1/ch20-02-multithreaded.md
- /book-1/ch20-03-graceful-shutdown-and-cleanup.md
- /book-1/foreword.md
- /book-1/title-page.md
- /book-2
- /book-2/README.md
- /book-2/SUMMARY.md
- /book-2/cli
- /book-2/cli/README.md
- /book-2/cli/build.md
- /book-2/cli/clean.md
- /book-2/cli/init.md
- /book-2/cli/serve.md
- /book-2/cli/test.md
- /book-2/cli/watch.md
- /book-2/continuous-integration.md
- /book-2/for_developers
- /book-2/for_developers/README.md
- /book-2/for_developers/backends.md
- /book-2/for_developers/preprocessors.md
- /book-2/format
- /book-2/format/README.md
- /book-2/format/config.md
- /book-2/format/mathjax.md
- /book-2/format/mdbook.md
- /book-2/format/summary.md
- /book-2/format/theme
- /book-2/format/theme/README.md
- /book-2/format/theme/editor.md
- /book-2/format/theme/index-hbs.md
- /book-2/format/theme/syntax-highlighting.md
- /book-2/misc
- /book-2/misc/contributors.md
- /book-3
- /book-3/SUMMARY.md
- /book-3/c1.md
- /book-3/c2.md
- /book-3/c3.md
- /book-3/c31.md
- /book-3/c32.md
- /book-3/c33.md
- /book-3/post1.md
- /book-3/pre1.md
- /book-3/pre2.md
- /xmark.toml
